packages/core/src/saml-applications/routes/utils.ts

```diff
@@ -28,8 +29,366 @@ import {
+  };
+
+  public getSignInUrl = async ({ scope, state }: { scope?: string; state?: string }) => {
+    const { authorizationEndpoint } = await this.fetchOidcConfig();
+
+    const queryParameters = new URLSearchParams({
+      [QueryKey.ClientId]: this.samlApplicationId,
+      [QueryKey.RedirectUri]: this.details.redirectUri,
+      [QueryKey.ResponseType]: 'code',
+      [QueryKey.Prompt]: Prompt.Login,
+    });
+
+    // TODO: get value of `scope` parameters according to setup in attribute mapping.
+    queryParameters.append(
+      QueryKey.Scope,
+      // For security reasons, DO NOT include the offline_access scope by default.
+      deduplicate([
+        ReservedScope.OpenId,
+        UserScope.Profile,
+        UserScope.Roles,
+        UserScope.Organizations,
+        UserScope.OrganizationRoles,
+        UserScope.CustomData,
+        UserScope.Identities,
+        ...(scope?.split(' ') ?? []),
+      ]).join(' ')
+    );
+
+    if (state) {
+      queryParameters.append(QueryKey.State, state);
+    }
+
+    return new URL(`${authorizationEndpoint}?${queryParameters.toString()}`);
+  };
+
+  protected getOidcConfig = async (): Promise<CamelCaseKeys<OidcConfigResponse>> => {
+    const oidcConfig = await tryThat(
+      async () => fetchOidcConfigRaw(this.issuer),
+      (error) => {
+        if (error instanceof ZodError) {
+          throw new RequestError({
+            code: 'oidc.invalid_request',
+            message: error.message,
+            error: error.flatten(),
+          });
+        }
+
+        throw error;
+      }
+    );
+
+    return oidcConfig;
+  };
+
+  protected exchangeAuthorizationCode = async ({ code }: { code: string }) => {
+    const { tokenEndpoint } = await this.fetchOidcConfig();
+    const result = await handleTokenExchange(tokenEndpoint, {
+      code,
+      clientId: this.samlApplicationId,
+      clientSecret: this.details.secret,
+      redirectUri: this.details.redirectUri,
+    });
+
+    if (!result.success) {
+      throw new RequestError({
+        code: 'oidc.invalid_token',
+        message: 'Invalid token response',
+      });
+    }
+
+    return camelcaseKeys(result.data);
+  };
+
+  protected async fetchOidcConfig() {
+    this.oidcConfig ||= await this.getOidcConfig();
+
+    return this.oidcConfig;
+  }
+
+  protected getUserInfo = async ({
+    accessToken,
+  }: {
+    accessToken: string;
+  }): Promise<IdTokenProfileStandardClaims & Record<string, unknown>> => {
+    const { userinfoEndpoint } = await this.fetchOidcConfig();
+    const body = await getRawUserInfoResponse(accessToken, userinfoEndpoint);
+    const result = idTokenProfileStandardClaimsGuard
+      .catchall(z.unknown())
+      .safeParse(parseJson(body));
+
+    if (!result.success) {
+      throw new RequestError({
+        code: 'oidc.invalid_request',
+        message: 'Invalid user info response',
+        details: result.error.flatten(),
+      });
+    }
+
+    return result.data;
+  };
```

Comment:
Per our offline discussion, can we create an OIDCConnector instance instead of rewriting all the OIDC methods here?

Reply:
True, double checked with OidcConnector class, we need to update the DB as well, see LOG-10689